### PR TITLE
Add JAX peekable iterator

### PR DESCRIPTION
**Pull request by awolant**
## Category:

**New feature**

## Description:

Adds `DALIGenericPeekableIterator` to JAX plugin. This iterator builds on top of regular DALI iterator for JAX. It adds `peek` and `peek_async` methods - to check next output without advancing the iterator. 
It also adds `element_spec` property - it returns object describing types and shapes of the outputs. 

It is compatible with Google CLU PeekableIterator.

### Affected modules and functionalities:

JAX plugin - I wanted the new iterator to be implemented in separate file. This way regular JAX iterator is not dependent on CLU. To do this I reorganized files in JAX plugin - made it multi file module with __init__.

### Key points relevant for the review:

- Is the structure of the new module ok?
- Is `peek_async` and related multi threaded code correct? 

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: 3605

### CODE REVIEW

**Comment by awolant** [comment]
!build

**Comment by awolant** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

File: dali/test/python/jax/test_peekable_iterator.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+shape = (1, 5)
+batch_shape = (batch_size, *shape[1:])
+
+
+def test_jax_peekable_iterator_peek():
+    # given
+    pipe = sequential_pipeline(batch_size, shape)
+
+    # when
+    iterator = DALIIteraor([pipe], ['data'], size=batch_size*100)
```
[the remark this answers is not in the document]
Comment (awolant) [reply dali/test/python/jax/test_peekable_iterator.py]: Done

File: dali/test/python/jax/test_peekable_iterator.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import jax.numpy as jnp
+from nvidia.dali.plugin.jax.clu import DALIGenericPeekableIterator as DALIIteraor
```
[the remark this answers is not in the document]
Comment (awolant) [reply dali/test/python/jax/test_peekable_iterator.py]: Done

File: dali/python/nvidia/dali/plugin/jax/iterator.py [hunk 3]
```diff
@@ -173,6 +173,7 @@ def __init__(
             prepare_first_batch=prepare_first_batch)
 
         self._first_batch = None
+        self._peek = None
```
Comment (stiepan) [inline dali/python/nvidia/dali/plugin/jax/iterator.py:176]: Do we need `_peek ` as member in the regular iterator?

Comment (awolant) [reply dali/python/nvidia/dali/plugin/jax/iterator.py]: Removed, leftover from some experiments.

**Review by stiepan — APPROVED** [review]
Looks good!

Just an ask for the comment in `peek_async` about the non-deterministic order when not waiting for the future when calling next or peek.

**Comment by awolant** [comment]
!build

**Pull request merged**